### DISCUSSION TITLE
Hierarchical relationship selection

### DISCUSSION THREAD

**Discussion by MrOtek598** [pocketbase/pocketbase#2769] [Q&A] [answered]
Hi,

I have these simple columns :

[slug]   [parent]
cat-3    cat-2
cat-2    cat-1
cat-1   N/A

I need to have the parents in hierarchy, like : cat-1/cat-2/cat-3.

Is there a way to do it in pb with one request ?

Thanks for your help

**Comment by ganigeorgiev** [comment]
> I need to have the parents in hierarchy, like : cat-1/cat-2/cat-3.

I'm not sure that I understand what is the desired output and how it would be used.
Do you want to generate a breadcrumb path from the posts or to be able to fetch a child with its parents?

To fetch a child with its parents you can use [expand](https://pocketbase.io/docs/expanding-relations/) (eg. `expand: 'parent.parent.parent'`), but note that it supports up to 6 level depth.

If you want a breadcrumb path for each post you can create a [Collection View](https://pocketbase.io/docs/collections/#view-collection) with recursive CTE that will concat the ids, something like (_I didn't test it_):

```sql
WITH breadcrumb AS 
(
  -- start from the root
  SELECT root.id, root.id AS path
  FROM posts root
  WHERE root.parent = ""

  UNION ALL

  -- concat the children
  SELECT child.id, (breadcrumb.path || '/' || child.id) AS path
  FROM posts child
  INNER JOIN breadcrumb ON breadcrumb.id = child.parent
)
SELECT
  posts.id,
  posts.slug,
  posts.parent,
  breadcrumb.path
FROM posts
INNER JOIN breadcrumb ON breadcrumb.id = posts.id 
```

**Reply by MrOtek598** [reply]
Thanks.

I tried with expand, but it's to heavy and limited for what I need.

I'm going to learn and explore about your way with recursive CTE, it seems to respond to what I need.

I just want to generate the path of my categories based on their hierarchy.

If a post is in these two categories : cat-2 and cat-3, and
If cat-3 have cat-2 as parent which have cat-1 as parent which finaly have no parent, I'm looking to simply get (in one request) these paths for the two categories of my post : cat-1/cat-2/cat-3 and cat-1/cat-2.

A screen of my simple collections :

![collections](https://github.com/pocketbase/pocketbase/assets/39051703/9237c1ad-a037-4a94-8c72-2f4cc196ebf4)
![posts](https://github.com/pocketbase/pocketbase/assets/39051703/91b7f99c-a112-415e-928c-1e200b076a9f)

----

I look a lot in the discussion area for answers to my questions, and it's fun how you say many time to not understand what we are trying to do (not always clear), but in the same time your answers respond to our needs !
Is it really you Gani or is it a member of the PB team dedicated to respond to us ? Because it's impressive how you respond well and quick ! I'm wondering how do you get the time??

Thanks a lot for your answers again.

**Reply by ganigeorgiev** [reply, marked as answer]
I don't think this can be achieved without a helper view collection. You can try:

1. Create a `categoriesView` with query like:
    ```sql
    With breadcrumb AS 
    (
      -- start from the root
      SELECT root.id, root.slug AS path
      FROM categories root
      WHERE root.parent = ""

      UNION ALL

      -- "merge" with with its children
      SELECT child.id, (breadcrumb.path || '/' || child.slug) AS path
      FROM categories child
      INNER JOIN breadcrumb ON breadcrumb.id = child.parent
    )
    SELECT
      categories.id,
      categories.id as category, -- used for referencing later via indirect expand
      breadcrumb.path
    FROM categories
    INNER JOIN breadcrumb ON breadcrumb.id = categories.id 
    ```
    This should produce something like:
    ```
    id1 | category-1.Id | "category-1"
    id2 | category-2.Id | "category-1/category-2"
    id3 | category-3.Id | "category-1/category-2/category-3"
    etc.
    ```

2. Then you can combine it with [indirect expand](https://pocketbase.io/docs/expanding-relations/#indirect-expand) for the `posts` categories:

    ```js
    await pb.collection("posts").getList(1, 30, {
      expand: "categories.categoriesView(category)"
    })
    ```

    The above will return 30 posts each with their category expanded and within each expanded category, another expand with their hierarchical path.

    You can limit the returned response fields using the `fields` query parameter.
    The current limitation is that there is no wildcard (`*`) and you have to list the exact fields you want to return, matching the json structure, eg.:

    ```js
    await pb.collection("posts").getList(1, 30, {
      expand: "categories.categoriesView(category)",
      fields: "id,slug,expand.categories.id,expand.categories.expand.categoriesView(category).path"
    })
    ```

**Reply by MrOtek598** [reply]
Thanks.

The query produce nothing in my case. It has to be me.
I will play with this query to find out what I'm doing wrong.

Thanks again, I learn a lot about pb capabilities through your answers.

**Comment by MrOtek598** [comment]
It works ! (In my "categories" collection, the parent field was a multi relational instead of a single relational field).
Thanks !
